feat: Add instruments service

YueChen-C opened the pull request:
inssues: #85
The test was successful only above 14

mykola-mokhnach commented on lib/instrument/headers.js:
```diff
@@ -0,0 +1,260 @@
+    messageHeader.writeUInt32LE(data.expectsReply, 28);
+    return messageHeader;
+  }
+  static parse (headerBuffer) {
```
Please keep at least one empty line indentation between class methods

mykola-mokhnach commented on lib/instrument/headers.js:
```diff
@@ -0,0 +1,260 @@
+    messageHeader.writeUInt32LE(data.flags, 0);
+    messageHeader.writeUInt32LE(data.auxLength, 4);
+    // eslint-disable-next-line no-undef
+    messageHeader.writeBigUInt64LE(BigInt(data.totalLength), 8);
```
should we explicitly cast this value? I would expect data.totalLength is already of the expected type because if it is not then this conversion does not make much sense - the value would be already cut to 52 bits

mykola-mokhnach commented on lib/instrument/headers.js:
```diff
@@ -0,0 +1,260 @@
+const DTX_MESSAGE_HEADER_MAGIC = 0x1F3D5B79;
+
+class DTXMessageHeader {
+  static build (data) {
```
could you please document all methods and data structures like it is done in other files?

mykola-mokhnach commented on lib/instrument/headers.js:
```diff
@@ -0,0 +1,260 @@
+    return {
+      flags: headerBuffer.readUInt32LE(0),
+      auxLength: headerBuffer.readUInt32LE(4),
+      totalLength: Number(headerBuffer.readBigInt64LE(8))
```
same as above - if we cast it to number then the further case to BigInt does not make much sense.

mykola-mokhnach commented on lib/instrument/headers.js:
```diff
@@ -0,0 +1,260 @@
+    // cursor += 8;
+    const data = [];
+    const length = Number(headerBuffer.readBigInt64LE(8));
+    cursor += 16;
```
what is the meaning of this magic number?

mykola-mokhnach commented on lib/instrument/headers.js:
```diff
@@ -0,0 +1,260 @@
+    cursor += 16;
+    while (cursor <= length) {
+      const m = headerBuffer.readUInt32LE(cursor);
+      if (m !== 0xa) {
```
this magic number could be a constant

mykola-mokhnach commented on lib/instrument/headers.js:
```diff
@@ -0,0 +1,260 @@
+    while (cursor <= length) {
+      const m = headerBuffer.readUInt32LE(cursor);
+      if (m !== 0xa) {
+        throw new Error('incorrect auxiliary magic');
```
It might be useful to also add the actual number we got into the error message for debugging purposes

mykola-mokhnach commented on lib/instrument/headers.js:
```diff
@@ -0,0 +1,260 @@
+}
+
+
+function InstrumentRPCParseError (data) {
```
I would rather follow the common pattern here and make this function to a class with a constructor

mykola-mokhnach commented on lib/instrument/headers.js:
```diff
@@ -0,0 +1,260 @@
+  }
+  getBytes () {
+    const out = Buffer.alloc(16);
+    out.writeUInt32LE(0x01F0, 0);
```
could be moved to a constant

mykola-mokhnach commented on lib/instrument/headers.js:
```diff
@@ -0,0 +1,260 @@
+  }
+}
+
+export { DTXMessageHeader, DTXMessagePayloadHeader, DTXMessageAuxBuffer, DTXMessageAux, DTXMessage, DTX_MESSAGE_PAYLOAD_HEADER_LENGTH, DTX_MESSAGE_HEADER_LENGTH, DTX_MESSAGE_HEADER_MAGIC};
```
Please split the long line into several shorter ones

mykola-mokhnach commented on lib/instrument/index.js:
```diff
@@ -0,0 +1,142 @@
+    this._undefinedCallback = event;
+    this._callbacks = new events.EventEmitter();
+    this._channelCallbacks = new events.EventEmitter();
+    this._replyQueues = defaultDict(function () {
```
I'd rather use an anonymous call here: `defaultDict(() => [])`

mykola-mokhnach commented on lib/instrument/index.js:
```diff
@@ -0,0 +1,142 @@
+    return channelID;
+  }
+  async waitReply (messageId, timeout = 10000) {
+    return await new B((resolve, reject) => {
```
we could use async methods imported from `asyncbox`, for example:

```
import { waitForCondition } from 'asyncbox';

...

return await waitForCondition(async () => {
        // returning true would exit the loop
        // exceptions thrown here are coming through
        return true;
      }, {waitMs: 10000, intervalMs: 500});
```

YueChen-C replied:
This's great

mykola-mokhnach commented on lib/instrument/index.js:
```diff
@@ -0,0 +1,142 @@
+    });
+  }
+
+  async call (channel, selector, ...auxiliaries) {
```
this method shadows the original https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/call, so I would rather rename it

mykola-mokhnach commented on lib/instrument/index.js:
```diff
@@ -0,0 +1,142 @@
+      return await this.waitReply(identifier, 10000);
+    }
+  }
+  _receiver (data) {
```
usually method names contain verbs. Can we make this one more readable?

mykola-mokhnach commented on lib/instrument/index.js:
```diff
@@ -0,0 +1,142 @@
+    }
+  }
+  _receiver (data) {
+    if (typeof (data.selector) == 'string' && data.selector.indexOf('_channelCanceled') !== -1) {
```
this condition could be simplified to `if (_.includes(data.selector, '_channelCanceled'))`. Also the magic string itself could be moved to constants

mykola-mokhnach commented on lib/instrument/index.js:
```diff
@@ -0,0 +1,142 @@
+    }
+    if (data.conversationIndex === 1) {
+      this._replyQueues[data.identifier].push(data);
+    } else if (this._channelCallbacks.listenerCount(2 ** 32 - data.channelCode) > 0) {
```
`2 ** 32` could be moved to constants

mykola-mokhnach commented on lib/instrument/index.js:
```diff
@@ -0,0 +1,142 @@
+      this._channelCallbacks.emit(2 ** 32 - data.channelCode, data);
+    } else {
+      const selector = data.selector;
+      if (typeof (selector) == 'string' && this._callbacks.listenerCount(selector) > 0) {
```
you could also use lodash's _.isString helper to check if a value is a string

mykola-mokhnach commented on lib/instrument/index.js:
```diff
@@ -0,0 +1,142 @@
+    reply._channel_code = data.channelCode;
+    reply._identifier = data.identifier;
+    reply._conversation_index = data.conversationIndex + 1;
+    reply._flags = 0x3;
```
could be moved to constants

mykola-mokhnach commented on lib/instrument/transformer/dtx-decode.js:
```diff
@@ -0,0 +1,70 @@
+    while (this.cursor < this.buffer.length) {
+      const magic = this._recv(4);
+      if (!magic) { return; }
+      this.cursor -= 4;
```
should the cursor be decreased or increased?

YueChen-C replied:
Need decreased

mykola-mokhnach commented on lib/instrument/transformer/dtx-decode.js:
```diff
@@ -0,0 +1,70 @@
+    this.buffer = Buffer.concat([this.buffer.slice(this.cursor, this.buffer.length), data]);
+    this.cursor = 0;
+    while (this.cursor < this.buffer.length) {
+      const magic = this._recv(4);
```
I assume 4 is the MAGIC_LEN

mykola-mokhnach commented on lib/instrument/transformer/dtx-decode.js:
```diff
@@ -0,0 +1,70 @@
+  }
+  _recv (length) {
+    if (this.buffer.length < this.cursor + length) {
+      return false;
```
I would rather return null, just to keep it consitent

mykola-mokhnach commented on lib/instrument/transformer/dtx-encode.js:
```diff
@@ -0,0 +1,32 @@
+  _encode (data) {
+    const {sync, channelID, selector, auxiliaries, identifier} = data;
+    const dtx = new DTXMessage();
+    dtx._identifier = identifier;
```
Methods/property names starting with underscore are considered private. Would it be possible to change the class constructor, so we could pass them as arguments upon instance creation?

mykola-mokhnach commented on lib/instrument/transformer/dtx-encode.js:
```diff
@@ -0,0 +1,32 @@
+    dtx._channelCode = channelID;
+    dtx._selector = selector;
+    dtx._expectsReply = sync;
+    if (auxiliaries.constructor.name === 'DTXMessageAuxBuffer') {
```
it would be more appropriate to use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/instanceof

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,389 @@
+}
+
+
+UNARCHIVE_CLASS_MAP.NSDictionary = function () {
```
we can simply init UNARCHIVE_CLASS_MAP upon declaration, e.g.:

```
const UNARCHIVE_CLASS_MAP = Object.freeze({
  NSDictionary: () => { UNARCHIVE_CLASS_MAP.decodeArchive = DictArchive; }
})

```

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,389 @@
+};
+
+UNARCHIVE_CLASS_MAP.NSDate = function () {
+  this.unix2apple_epoch_delta = 978307200.0;
```
can we move the magic number to constants?

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,389 @@
+
+  unpackArchiveHeader () {
+    const plist = plistlib.parseBuffer(this.input)[0];
+    if (plist.$archiver !== 'NSKeyedArchiver') {
```
Can we move magic values to constants?

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,389 @@
+    const top = plist.$top;
+    const topUID = top.root;
+    if (!topUID) {
+      log.error(`top object did not have a UID! dump: ${top}`);
```
if `top` is a plain object then something like `[Object object]` might be printed to the console. We might consider using `JSON.stringify()` to see what it actually contains

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,389 @@
+  }
+  decodeKey (obj, key) {
+    const val = obj[key];
+    if (typeof val === 'object' && val.UID) {
```
this might fail val is `null`. `if (!_.isNil(val?.UID))` would be safer

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,389 @@
+    if (typeof val === 'object' && val.UID) {
+      return this.decodeObject(val);
+    }
+    return val;
```
this method could use a ternary operator

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,389 @@
+    const rawObj = this.objects[index.UID];
+    this.unpackedUids[index.UID] = CYCLE_TOKEN;
+
+    if (typeof rawObj !== 'object' || !rawObj.$class) {
```
if (!_.isNil(rawObj?.$class))

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,389 @@
+      return rawObj;
+    }
+    const classUid = rawObj.$class;
+    if (!classUid) {
```
this check is already made above

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,389 @@
+    }
+    const index = new plistlib.UID(this.objects.length);
+    const objType = typeof obj;
+    if (['number', 'string', 'boolean', 'UID'].indexOf(objType) !== -1) {
```
- this list could be extracted to a constant
- consider using https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes instead of indexOf

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,389 @@
+        return this.encodeDict(obj, archiveObj);
+      }
+    }
+
```
should we throw an exception if none of the above matched?

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,389 @@
+  }
+  encodeArray (objs, archiveObj) {
+    archiveObj.$class = this.uidForArchiver('NSArray');
+    archiveObj['NS.objects'] = objs.map((data) => this.archive(data));
```
could be simplified to objs.map(this.archive.bind(this))

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,389 @@
+    let keys = [];
+    let vals = [];
+
+    for (const k in obj) {
```
could be simplified to 
```
const keys = _.keys(obj).map(this.acrhive.bind(this));
const vals = _.values(obj).map(this.acrhive.bind(this));
```

mykola-mokhnach commented on lib/services.js:
```diff
@@ -80,6 +82,20 @@ async function startHouseArrestService (udid, opts = {}) {
 
+async function startInstrumentService (udid, opts = {}) {
+  const osVersion = opts.osVersion || await getOSVersion(udid, opts.socket);
+  if (Number(osVersion.split('.')[0]) < 14) {
```
use `parseInt(x, 10)` to convert string to an integer
`14` could be moved to constants

mykola-mokhnach commented on lib/services.js:
```diff
@@ -80,6 +82,20 @@ async function startHouseArrestService (udid, opts = {}) {
+  if (Number(osVersion.split('.')[0]) < 14) {
+    const socket = await startService(udid, INSTRUMENT_SERVICE_NAME, opts.socket);
+    // # after the ssl protocol is successfully handshake, I need to close the ssl protocol channel and use text transmission
+    // socket.ssl =  null
```
for now it would be ok to throw an exception here that we don't support iOS below 14

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,389 @@
+import plistlib from 'bplist-parser';
+import bplistCreate from 'bplist-creator';
+import log from 'crypto';
+import {parse as uuidParse, stringify as uuidStringify} from 'uuid';
```
this must be included to package.json in order to be used

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,389 @@
+import plistlib from 'bplist-parser';
+import bplistCreate from 'bplist-creator';
+import log from 'crypto';
```
why log is imported from `crypto`? Simply follow examples from other modules

mykola-mokhnach commented on lib/instrument/headers.js:
```diff
@@ -0,0 +1,260 @@
+import {archive, unarchive} from './transformer/nskeyed';
+import plistlib from 'bplist-parser';
```
npm modules cannot be used unless they are defined as depencies in package.json
Also, all plist-related stuff could be found in the `@appium/support` lib. Check https://github.com/appium/appium/blob/2.0/packages/support/test/plist-specs.js and other examples from this module for more details

mykola-mokhnach commented on lib/instrument/headers.js:
```diff
@@ -0,0 +1,260 @@
+    return this._selector;
+  }
+  get conversationIndex () {
+    return this._messageHeader.conversationIndex;
```
such getter might throw exceptions if the corresponding properties are undefined. It would be safer to use the following format: `return this._messageHeader?.conversationIndex;`

mykola-mokhnach commented:
It would be nice to have some unit tests. At least these utility transformation classes could easily be checked using a synthetic data

KazuCocoa commented:
thank you! Let me play with this branch for ios 14-. (hopefully this weekend)

YueChen-C commented:
@mykola-mokhnach 
Thank you so much for taking the time to review. I modified the code:
1. The instrument is running under version 14
2. According to your suggestion, I modified some code
3. Add test cases
4. Add document

KazuCocoa commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,378 @@
+      this.objects.push(obj);
+      return index;
+    }
+    let archiveObj = {};
```
can be `const`?

KazuCocoa reviewed (commented):
Great, thank you... will take a look tomorrow again.

mykola-mokhnach commented on lib/instrument/headers.js:
```diff
@@ -0,0 +1,339 @@
+      if (typeof arg === 'number') {
+        if (Math.abs(arg) < INT32) {
+          messageAux.appendInt(arg);
+        } else if (Math.abs(arg) < INT64) {
```
javascript integer numbers can never reach the size greater of [Number.MAX_SAFE_INTEGER](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER). All numbers larger that that would be cut off unless these are stored as BigInt

mykola-mokhnach commented on lib/instrument/headers.js:
```diff
@@ -0,0 +1,339 @@
+        } else if (Math.abs(arg) < INT64) {
+          messageAux.appendLong(arg);
+        } else {
+          throw new Error('num too large');
```
this would probably never happen unless we have a float number. Also could the error message be improved?

mykola-mokhnach commented on lib/instrument/headers.js:
```diff
@@ -0,0 +1,339 @@
+
+  static parse (headerBuffer) {
+    let cursor = 0;
+    // // magic not find the rule
```
do we still need the commented code?

mykola-mokhnach commented on lib/instrument/headers.js:
```diff
@@ -0,0 +1,339 @@
+const DTX_MESSAGE_HEADER_MAGIC = 0x1F3D5B79;
+const DTX_MESSAGE_HEADER_MAGIC_LEN = 4;
+const DTX_MESSAGE_AUX_HEADER = 0x01F0;
+// const DTX_MESSAGE_AUX_MAGIC = 0x1df0;
```
do we need the commented value?

mykola-mokhnach commented on lib/instrument/headers.js:
```diff
@@ -0,0 +1,339 @@
+        case AUX_TYPES.BigInt64LE:
+          data.push(headerBuffer.readBigInt64LE(cursor));
+          cursor += 8;
+          break;
```
should we gave a `default` branch here?

mykola-mokhnach commented on lib/instrument/headers.js:
```diff
@@ -0,0 +1,339 @@
+    buf.writeUInt32LE(3, 4);
+    buf.writeUInt32LE(value, 8);
+    this._buf = Buffer.concat([this._buf, buf]);
+
```
the empty line at the end of each method is redundant

mykola-mokhnach commented on lib/instrument/headers.js:
```diff
@@ -0,0 +1,339 @@
+    buf.writeUInt32LE(10, 0);
+    buf.writeUInt32LE(4, 4);
+    // eslint-disable-next-line no-undef
+    buf.writeBigUInt64LE(BigInt(value), 8);
```
this won't really work if the incoming value is not of type BigInt already

YueChen-C replied:
I don't know how to write uInt64 in nodejs Buffer. This approach seems to work

YueChen-C replied:
Using this method requires converting int to bigint , or what should i do？

mykola-mokhnach replied:
like I said above this works, but actually does not do anything. we should make sure that the incoming value has the correct type already (e.g. make a BigInt when it is read from Buffer). Check https://nodejs.org/api/buffer.html#bufreadbigint64leoffset for more details

YueChen-C replied:
I know what you mean, but the totalLength is generated by the DTXMessageAuxBuffer and selectorBuffer lengths. convert bigint should be required.
```
totalLength: aux.length + sel.length,
```
I will modify other code about bigint

mykola-mokhnach replied:
Then you probably need to do some more magic based on the actual variable length (e.g. if it is more than 4 bytes that use bigint otherwise a vanilla number). Check https://stackoverflow.com/questions/52595180/how-to-read-write-a-bigint-from-buffer-in-node-js-10 for more details on how buffer values could be properly packed/unpacked into bigint

mykola-mokhnach commented on lib/instrument/headers.js:
```diff
@@ -0,0 +1,339 @@
+    this._buf.writeUInt32LE(10, 0);
+    this._buf.writeUInt32LE(6, 4);
+    // eslint-disable-next-line no-undef
+    this._buf.writeBigInt64LE(BigInt(value), 8);
```
same as above

mykola-mokhnach commented on lib/instrument/headers.js:
```diff
@@ -0,0 +1,339 @@
+  }
+
+  /**
+   * DTXMessage Buffer: DTXMessageHeader + PayloadHeader + DTXMessageAuxBuffer(nullable) + selector(nullable)
```
👍

mykola-mokhnach commented on lib/instrument/headers.js:
```diff
@@ -0,0 +1,339 @@
+  }
+
+  /**
+   * @param headerBuf
```
Would it be possible to also describe argument types and purpose like in other modules?

mykola-mokhnach commented on lib/instrument/index.js:
```diff
@@ -0,0 +1,176 @@
+import _ from 'lodash';
+import {waitForCondition} from 'asyncbox';
+const CHECK_FREQ_MS = 500;
+const WAIT_REPLY_TIME = 10000;
```
I usually add unit suffix to each constant name to avoid confusion, like it was done in the previous one (`_MS`)

mykola-mokhnach commented on lib/instrument/index.js:
```diff
@@ -0,0 +1,176 @@
+  /**
+   * Create a service channel, data transmission of the service will use this channelCode
+   * @param {String} channel  the parameter is instruments Service e.g: SERVICE.DEVICE_INFO
+   * @returns {Promise<number|*>}
```
would it be possible to also describe the purpose of the returned result?

mykola-mokhnach commented on lib/instrument/index.js:
```diff
@@ -0,0 +1,176 @@
+   * In general, we call this method to start the instrument dtx service
+   * @param {string} channel instrument dtx service e.g: SERVICE.DEVICE_INFO
+   * @param {string} selector service method
+   * @param auxiliaries parameters required by selector
```
the type would be `{...any}`

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,378 @@
+const CYCLE_TOKEN = 'CycleToken';
+const PRIMITIVE_TYPES = ['number', 'string', 'boolean', 'UID'];
+const NSKEYEDARCHIVER = 'NSKeyedArchiver';
+const UNIX2APPLE_EPOCH_DELTA = 978307200.0;
```
can we add unit name suffix to the constant name?

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,378 @@
+
+UNARCHIVE_CLASS_MAP.ExceptionArchive = function () {
+  this.decodeArchive = function (archive) {
+    return {'$class': 'NSException', 'reason': archive.decode('NS.reason'), 'userinfo': archive.decode('userinfo'), 'name': archive.decode('NS.name')};
```
can we split the object definition into multiple lines?

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,378 @@
+  }
+
+  archive (obj) {
+    if (typeof obj === 'undefined') {
```
you could also use _.isUndefined lodash helper for this purpose

mykola-mokhnach commented on lib/ssl-helper.js:
```diff
@@ -1,7 +1,9 @@
+import { waitForCondition } from 'asyncbox';
 
 const TLS_VERSION = 'TLSv1_method';
+const HANDSHAKE_TIMEOUT = 10000;
```
it would make sense to add the unit name suffix to the const name

mykola-mokhnach commented on lib/ssl-helper.js:
```diff
@@ -13,4 +15,46 @@ function upgradeToSSL (socket, key, cert) {
+
+  // stop receiving data after successful handshake
+  sslSocket.on('secureConnect', () => {
+    sslSocket._handle.readStop();
```
nice find

mykola-mokhnach commented on lib/ssl-helper.js:
```diff
@@ -13,4 +15,46 @@ function upgradeToSSL (socket, key, cert) {
+  await waitForCondition(() => isSecureConnect, {
+    waitMs: HANDSHAKE_TIMEOUT,
+    intervalMs: 500,
+    error: 'ssl handshake error'}).catch((err) => {
```
Please use try/catch instead. .catch only makes sense if await is not used

mykola-mokhnach commented on lib/ssl-helper.js:
```diff
@@ -13,4 +15,46 @@ function upgradeToSSL (socket, key, cert) {
+    isSecureConnect = true;
+  });
+
+  await waitForCondition(() => isSecureConnect, {
```
probably for this particular case the usage is not optimal, because it would still make 500ms delay after the connection is done. I wuld try something like that:

```
await new B((resolve, reject) => {
  sslSocket.once('secureConnect', () => {
    return resolve();
  });
  
  setTimeout(HANDSHAKE_TIMEOUT, () => {
      if (!sslSocket.destroyed) {
       sslSocket.end();
     }
     return reject(new Error('ssl handshake error'));
  });
});

return net.Socket({fd: socket._handle.fd});
```

mykola-mokhnach commented on package.json:
```diff
@@ -34,7 +34,11 @@
     "semver": "^7.0.0",
-    "source-map-support": "^0.x"
+    "source-map-support": "^0.x",
+    "asyncbox": "^2.9.2",
```
Please sort imports alphabetically

mykola-mokhnach commented on package.json:
```diff
@@ -34,7 +34,11 @@
+    "source-map-support": "^0.x",
+    "asyncbox": "^2.9.2",
+    "uuid": "^8.3.2",
+    "bplist-creator": "0.1.0",
```
so we really want to stick to particular versions of these modules? We usually allow everything below the next major version, e.g.: `"bplist-creator": "^0.x",`

mykola-mokhnach commented on test/instrument/instrument-service-specs.js:
```diff
@@ -0,0 +1,56 @@
+      }, 2000);
+    });
+    data.should.to.deep.equal([0, 51, 59, 60]);
+
```
extra empty line

mykola-mokhnach commented on test/instrument/nskeyed-specs.js:
```diff
@@ -0,0 +1,29 @@
+
+describe('NSKeyedArchive', function () {
+
+  it('should NSKeyedArchive and NSKeyedUnArchive', function () {
```
can we have better scenario name?

mykola-mokhnach commented on test/instrument/instrument-service-specs.js:
```diff
@@ -0,0 +1,56 @@
+import chai from 'chai';
```
👍 
I like the tests

KazuCocoa replied:
me too👍

KazuCocoa commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,378 @@
+/**
+ * Capable of unpacking an archived object tree in the NSKeyedArchive format.
+ * Apple's implementation can be found here:
+ * https://github.com/apple/swift-corelibs-foundation/blob/master/Foundation/NSKeyedUnarchiver.swift
```
it seems like they changed the branch name: https://github.com/apple/swift-corelibs-foundation/blob/main/Sources/Foundation/NSKeyedUnarchiver.swift
(other comment also should update `master` to `main`)

mykola-mokhnach commented on test/fixtures/index.js:
```diff
@@ -26,7 +26,10 @@ const fixtures = {
+  AFC_FILE_INFO_RESPONSE: 'afcFileInfoResponse',
+  INSTRUMENTS_LAUNCH_APP: 'instrumentsLaunchApp',
+  INSTRUMENTS_FPS: 'instrumentsFps',
+
```
this empty line is not needed

YueChen-C commented:
I revised the above question @mykola-mokhnach

YueChen-C commented:
I revised the above question @mykola-mokhnach

mykola-mokhnach commented:
Great job @YueChen-C, thanks
I've added a few more comments, but in general the PR looks good

YueChen-C commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,310 @@
+ * decode and encode Archive of currently known data formats
+ * @type {Readonly}
+ */
+const UNARCHIVE_CLASS_MAP = Object.freeze({
```
@mykola-mokhnach could you review the code for me. I learned those websites and refactored here

mykola-mokhnach replied:
this would probably not work as expected. If you just want to map the particular class then the following needs to be done:

- remove the decodeArchive and encodeArchive global module vars
- `DTSysmonTapMessage: DTTapMessagePlist` should be ok
- Create a base class : `BaseMessagePlist` with two methods `decodeArchive` and `encodeArchive` which implement the default behaviour or throw an exception
- Inherit other classes from it and override the above methods
- The code to extract the necessary method would look like:

```js
    const classUid = rawObj.$class;
    const instance = this.classForUid(classUid)();
    const decodedObj = instance.decodeArchive(new ArchivedObject(rawObj, this));
```

mykola-mokhnach replied:
Also adding a simple unit test on these method would be helpful to test the actual encoding/decoding behavior

YueChen-C replied:
In the beginning I planned to do it this way.
But when I study javascript, The document shows that you can simply create a class
```javascript
UNARCHIVE_CLASS_MAP.NSDictionary = function () {
  this.decodeArchive = DictArchive;
};

new UNARCHIVE_CLASS_MAP['NSDictionary']().decodeArchive()
```
I seem to be going in the wrong direction

YueChen-C replied:
This looks almost the same as python. Let me revise it again

mykola-mokhnach replied:
yes, I'm also more like a Python person myself. Try to make it simple/easy to read. It is only necessary to use a special javascript dark magic where it is not possible to find other ways.

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,310 @@
+
+function ListArchive (archive) {
+  const uids = archive.decode('NS.objects');
+  const _list = [];
```
could be simplified to `return uids.map((uid) => archive.decodeIndex(uid))`

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,310 @@
+  }
+
+  encode (val) {
+    if (NOT_ENCODE_TYPES.includes(typeof val)) {
```
NOT_ENCODE_TYPES -> NOT_ENCODABLE_TYPES

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,310 @@
+        return this.encodeDict(obj, archiveObj);
+      }
+    } else {
+      throw Error(`Unable to encode types: ${obj}`);
```
would it make more sense to print `typeof obj`?

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,310 @@
+  /**
+   * use the UNARCHIVE_CLASS_MAP to find the unarchiving delegate of a uid
+   */
+  classForUid (index) {
```
this method could probably return an instance instead of a class, since we anyway call `new` on its result afterwards

YueChen-C commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,431 @@
+/**
+ * decode and encode Archive of currently known data formats
+ */
+const UNARCHIVE_CLASS_MAP = {
```
This looks so much better, could you help me check it again @mykola-mokhnach

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,431 @@
+ * @param {Archive} archiver
+ * @constructor
+ */
+function ArchivingObject (object, archiver) {
```
I would rather make it to a class

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,431 @@
+ * @param {Unarchive} unarchiver
+ * @constructor
+ */
+function ArchivedObject (object, unarchiver) {
```
I would rather make it to a class

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,431 @@
+import {parse as uuidParse, stringify as uuidStringify} from 'uuid';
+import _ from 'lodash';
+
+const NSKeyedArchiveVersion = 100_000;
```
constants should only include uppercase letters

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,431 @@
+      return rawObj;
+    }
+    const klass = this.classForUid(rawObj.$class);
+    const klassObj = new klass().decodeArchive(new ArchivedObject(rawObj, this));
```
some classes, like NSUUID require constructor arguments to be provided. How do we handle this situation here?

YueChen-C replied:
We do not need constructor arguments in the decoding process. constructor arguments required only when encoding process
Refer to nskeyed-specs.js unit test case

mykola-mokhnach replied:
Could you please help me to find the place where arguments are passed to the corresponding constructors for further encoding?

YueChen-C replied:
The instance has been created before encode . Like this
``` javascript
const data = {'NSURL': new NSURL(null, 'file://PlugIns/WebDriverAgentRunner.xctest')};
const archiveData = archive(data);
```
After that NSURL will be encoded in this code. `obj` is an instance of the NSURL
```javascript
if (objName in UNARCHIVE_CLASS_MAP) {
  archiveObj.$class = this.uidForArchiver(objName);
  obj.encodeArchive(obj, new ArchivingObject(archiveObj, this));
```

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,431 @@
+      return this.encodeSet(obj, archiveObj);
+    } else if (obj instanceof Object) {
+      const objName = obj.constructor.name;
+      if (objName in UNARCHIVE_CLASS_MAP) {
```
can we rely on this condition? not all keys in this dictionary match to class names

YueChen-C replied:
Only special class/instance are useful, such as NSURL, NSUUID, NSDate .And this class must also have the encodeArchive method

mykola-mokhnach replied:
makes sense. Perhaps it would be easier to read this could if we had the above explanation as a comment there

mykola-mokhnach commented on lib/instrument/transformer/nskeyed.js:
```diff
@@ -0,0 +1,431 @@
+ * @param {BaseArchiveHandler} subClass inherit from BaseArchiveHandler class
+ */
+function updateNSKeyedArchiveClass (name, subClass) {
+  if (!(Object.getPrototypeOf(subClass) === BaseArchiveHandler)) {
```
In languages with dynamic typing I would prefer to rather use duck-style check to explicit type checks. For example, this condition could check if the given class/instance have encodeArchive/decodeArchive methods

YueChen-C replied:
Like this?
```
if (!_.isFunction(subClass.prototype.decodeArchive) && !_.isFunction(subClass.prototype.encodeArchive))
```

mykola-mokhnach replied:
yes. In strictly typed languages one would use an interface for such purprose

mykola-mokhnach replied:
you could also use `subClass.prototype?.decodeArchive` to make the check safer

mykola-mokhnach reviewed (approved):
LGTM

@KazuCocoa could you please take another look at this PR when you have time?

KazuCocoa commented:
Yes, will take a look tomorrow or this weekend.

KazuCocoa reviewed (approved):
seems good to me..
